extensions/binding/org.eclipse.smarthome.binding.openweathermap/src/main/java/org/eclipse/smarthome/binding/openweathermap/internal/connection/OpenWeatherMapConnection.java

```diff
@@ -0,0 +1,232 @@
+        params.put(PARAM_LON, locationConfig.getLongitude().toString());
+
+        // Multilingual support (see https://openweathermap.org/current#multi)
+        String language = StringUtils.trimToEmpty(apiConfig.getLanguage());
+        if (!language.isEmpty()) {
+            params.put(PARAM_LANG, language.toLowerCase());
+        }
+        return params;
+    }
+
+    private String buildURL(String url, Map<String, String> requestParams) {
+        return requestParams.keySet().stream().map(key -> key + "=" + encodeParam(requestParams.get(key)))
+                .collect(joining("&", url + "?", StringUtils.EMPTY));
+    }
+
+    private String encodeParam(String value) {
+        try {
+            return URLEncoder.encode(value, UTF_8_ENCODING);
+        } catch (UnsupportedEncodingException e) {
+            logger.debug("UnsupportedEncodingException occurred during execution: {}", e.getLocalizedMessage(), e);
+            return StringUtils.EMPTY;
+        }
+    }
+
+    private String getResponseFromCache(String url)
+            throws OpenWeatherMapCommunicationException, OpenWeatherMapConfigurationException {
+        return cache.putIfAbsentAndGet(url, () -> getResponse(url));
+    }
+
+    private String getResponse(String url)
+            throws OpenWeatherMapCommunicationException, OpenWeatherMapConfigurationException {
+        try {
+            logger.trace("OpenWeatherMap request: URL = '{}'", url);
+            ContentResponse contentResponse = httpClient.newRequest(url).method(GET).timeout(10, TimeUnit.SECONDS)
+                    .send();
+            int httpStatus = contentResponse.getStatus();
+            String content = contentResponse.getContentAsString();
+            String errorMessage = StringUtils.EMPTY;
+            logger.trace("OpenWeatherMap response: status = {}, content = '{}'", httpStatus, content);
+            switch (httpStatus) {
+                case OK_200:
+                    return content;
+                case BAD_REQUEST_400:
+                case UNAUTHORIZED_401:
+                    errorMessage = getErrorMessage(content);
+                    logger.debug("OpenWeatherMap server responded with status code {}: {}", httpStatus, errorMessage);
+                    throw new OpenWeatherMapConfigurationException(errorMessage);
+                case TOO_MANY_REQUESTS_429:
+                    // TODO disable refresh job temporarily (see https://openweathermap.org/appid#Accesslimitation)
+                default:
+                    errorMessage = getErrorMessage(content);
+                    logger.debug("OpenWeatherMap server responded with status code {}: {}", httpStatus, errorMessage);
+                    throw new OpenWeatherMapCommunicationException(errorMessage);
+            }
+        } catch (ExecutionException e) {
+            String errorMessage = e.getLocalizedMessage();
+            if (errorMessage.startsWith("org.eclipse.jetty.client.HttpResponseException")) {
```

Comment:
Does this have to be a string matching here? What is different about the 401 handling that you have in line 199 already?

Reply:
This is one of the most ugliest part I discovered during the implementation of the binding. Funny thing is once I try to request the OMW API with an invalid a API key I do not get a `401 Unauthorized` status code but Jetty internally fails with an `org.eclipse.jetty.client.HttpResponseException` -> `ExecutionException`. The same request in a browser fails in the normal way returning a `401 Unauthorized` status code. If you have any idea why or how to solve it in a better way your help is very much appreciated. Try it yourself:

https://api.openweathermap.org/data/2.5/forecast/daily?appid=12345&cnt=6&lon=25.782403&units=metric&lang=de&lat=-80.264563

Reply:
What is the `cause` of the `ExecutionException`?

Reply:
Here is a stack trace for the failing request:
```
2018-09-25 13:54:30.494 [TRACE] [.connection.OpenWeatherMapConnection] - Exception occurred during execution: org.eclipse.jetty.client.HttpResponseException: HTTP protocol violation: Authentication challenge without WWW-Authenticate header
java.util.concurrent.ExecutionException: org.eclipse.jetty.client.HttpResponseException: HTTP protocol violation: Authentication challenge without WWW-Authenticate header
	at org.eclipse.jetty.client.util.FutureResponseListener.getResult(FutureResponseListener.java:118) ~[?:?]
	at org.eclipse.jetty.client.util.FutureResponseListener.get(FutureResponseListener.java:101) ~[?:?]
	at org.eclipse.jetty.client.HttpRequest.send(HttpRequest.java:684) ~[?:?]
	at org.eclipse.smarthome.binding.openweathermap.internal.connection.OpenWeatherMapConnection.getResponse(OpenWeatherMapConnection.java:190) ~[?:?]
	at org.eclipse.smarthome.binding.openweathermap.internal.connection.OpenWeatherMapConnection.lambda$2(OpenWeatherMapConnection.java:182) ~[?:?]
	at org.eclipse.smarthome.core.cache.ExpiringCache.refreshValue(ExpiringCache.java:91) ~[?:?]
	at org.eclipse.smarthome.core.cache.ExpiringCache.getValue(ExpiringCache.java:61) ~[?:?]
	at org.eclipse.smarthome.core.cache.ExpiringCacheMap.get(ExpiringCacheMap.java:202) ~[?:?]
	at org.eclipse.smarthome.core.cache.ExpiringCacheMap.putIfAbsentAndGet(ExpiringCacheMap.java:133) ~[?:?]
	at org.eclipse.smarthome.core.cache.ExpiringCacheMap.putIfAbsentAndGet(ExpiringCacheMap.java:118) ~[?:?]
	at org.eclipse.smarthome.binding.openweathermap.internal.connection.OpenWeatherMapConnection.getResponseFromCache(OpenWeatherMapConnection.java:182) ~[?:?]
	at org.eclipse.smarthome.binding.openweathermap.internal.connection.OpenWeatherMapConnection.getDailyForecastData(OpenWeatherMapConnection.java:124) ~[?:?]
	at org.eclipse.smarthome.binding.openweathermap.internal.handler.OpenWeatherMapDailyForecastHandler.requestData(OpenWeatherMapDailyForecastHandler.java:56) ~[?:?]
	at org.eclipse.smarthome.binding.openweathermap.internal.handler.AbstractOpenWeatherMapHandler.updateData(AbstractOpenWeatherMapHandler.java:125) ~[?:?]
	at org.eclipse.smarthome.binding.openweathermap.internal.handler.OpenWeatherMapAPIHandler.updateThing(OpenWeatherMapAPIHandler.java:154) ~[?:?]
	at org.eclipse.smarthome.binding.openweathermap.internal.handler.OpenWeatherMapAPIHandler.lambda$2(OpenWeatherMapAPIHandler.java:121) ~[?:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
	at java.lang.Thread.run(Thread.java:748) [?:?]
Caused by: org.eclipse.jetty.client.HttpResponseException: HTTP protocol violation: Authentication challenge without WWW-Authenticate header
	at org.eclipse.jetty.client.AuthenticationProtocolHandler$AuthenticationListener.onComplete(AuthenticationProtocolHandler.java:197) ~[?:?]
	at org.eclipse.jetty.client.ResponseNotifier.notifyComplete(ResponseNotifier.java:216) ~[?:?]
	at org.eclipse.jetty.client.ResponseNotifier.notifyComplete(ResponseNotifier.java:208) ~[?:?]
	at org.eclipse.jetty.client.HttpReceiver.terminateResponse(HttpReceiver.java:470) ~[?:?]
	at org.eclipse.jetty.client.HttpReceiver.responseSuccess(HttpReceiver.java:416) ~[?:?]
	at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.messageComplete(HttpReceiverOverHTTP.java:316) ~[?:?]
	at org.eclipse.jetty.http.HttpParser.handleContentMessage(HttpParser.java:605) ~[?:?]
	at org.eclipse.jetty.http.HttpParser.parseContent(HttpParser.java:1675) ~[?:?]
	at org.eclipse.jetty.http.HttpParser.parseNext(HttpParser.java:1523) ~[?:?]
	at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.parse(HttpReceiverOverHTTP.java:172) ~[?:?]
	at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.process(HttpReceiverOverHTTP.java:135) ~[?:?]
	at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.receive(HttpReceiverOverHTTP.java:73) ~[?:?]
	at org.eclipse.jetty.client.http.HttpChannelOverHTTP.receive(HttpChannelOverHTTP.java:133) ~[?:?]
	at org.eclipse.jetty.client.http.HttpConnectionOverHTTP.onFillable(HttpConnectionOverHTTP.java:155) ~[?:?]
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:281) ~[?:?]
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:102) ~[?:?]
	at org.eclipse.jetty.io.ssl.SslConnection.onFillable(SslConnection.java:291) ~[?:?]
	at org.eclipse.jetty.io.ssl.SslConnection$3.succeeded(SslConnection.java:151) ~[?:?]
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:102) ~[?:?]
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:118) ~[?:?]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:333) ~[?:?]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:310) ~[?:?]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:168) ~[?:?]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:126) ~[?:?]
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:366) ~[?:?]
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:762) ~[?:?]
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:680) ~[?:?]
	... 1 more
```

Reply:
Just to make clear: The same request with a valid API key works like a charm.

Reply:
My point has been: Shouldn't we check the e.g. the cause using `instanceof` instead of check the message string?

Reply:
Seems as if the service does not respond with a proper HTTP authentication response.

But couldn't you check the `cause` for being an instance of `HttpResponseException` instead of doing string comparison?

Reply:
I get it. You mean we should replace
```
if (errorMessage.startsWith("org.eclipse.jetty.client.HttpResponseException")) {
...
```
with
```
if (e.getCause() instanceof HttpResponseException) {
...
```
That is nice and should work too.